### DISCUSSION TITLE
visual improvement?

### DISCUSSION THREAD

**Discussion by JonaTh4nX** [mkkellogg/GaussianSplats3D#174] [Q&A]
Hi there ! 🙂 

Here is one of my capture using Splatfacto training method with 30K ite .

In the image on the left, this is a capture in Unity showcasing the rendering quality achieved through the plugin developed by [Aras-P](https://github.com/aras-p/UnityGaussianSplatting) , which is nearly perfect regarding the visual quality. 

On the image to the right, this is a capture from the Mark Kellogg's viewer ( and it look roughly the same using the viewer of [Antimatter](https://github.com/antimatter15/splat)).

Take a look at the wall on the right, the sky, the shadow of the roof on the white wall to the left, or the car's bodywork... Why is there such a visual difference? Is there any room for improvement on this?
<img width="1351" alt="side(12)" src="https://github.com/mkkellogg/GaussianSplats3D/assets/159079347/3819b4df-94ce-4cb3-a4e6-009942ed018e">

**Comment by mkkellogg** [comment]
Others have had similar issues and I'm not exactly sure what the root cause is. Would you be able to share your capture so I can do some debugging on my end? Also, have you tried viewing your capture in both/either PlayCanvas' SuperSplat editor or Polycam's viewer? I'd be interested to see how it looks in those viewers.

**Reply by JonaTh4nX** [reply]
Regarding the visual I provided, it seems to come from the training method. I'm working on it ....

Nevertheless, when we look more closely, there are visual differences between the viewers. Unity offers by far the most beautiful visual rendering, while Polycam is not far behind, and from what I have observed so far, it seems to be the web viewer that comes closest to the visual quality of Unity. Yours remains very competitive because it is much smoother in terms of navigation and has many other advantages! Do you think your viewer can reach the visual quality of Polycam?

Here are some screenshots to illustrate all this.

![Comparatif-2](https://github.com/mkkellogg/GaussianSplats3D/assets/159079347/942363fc-cae1-4548-9a1e-f0b7eeb0318f)
![Comparatif-1](https://github.com/mkkellogg/GaussianSplats3D/assets/159079347/79420db2-b0ca-4df8-9d19-42612735ce3c)

**Reply by JR3nard** [reply]
Hello all, 

I've noticed that we have a similar issue. With the SIBR local viewer the result is AWESOME but when we use your web viewer, the quality seems to be bellow. 

I've see this thread: https://github.com/graphdeco-inria/gaussian-splatting/issues/455 
![image](https://github.com/mkkellogg/GaussianSplats3D/assets/165813249/ce84134a-987c-4017-b09e-d2cee69c03f5)

It's seems to be related to the spherical harmonics and the renderer/diff-gaussian-rasterization

@mkkellogg have you the possibility to implement the support for different degree of spherical harmonics ? With the SIBR local viewer, it support sh_degree = 3. It can have a big impact on the quality.

**Comment by mkkellogg** [comment]
So the discrepancy seems to be mostly due to the lack of support for spherical harmonics in my renderer. When I load up the models in Aras Pranckevičius' Unity-based viewer (https://github.com/aras-p/UnityGaussianSplatting), and edit the code to disable the spherical harmonics support, the output looks much, much closer to mine.

It looks like I need to prioritize support for spherical harmonics :) The big reason I have avoided it is because of the large memory requirement, but maybe some sort of minimal-loss compression scheme could make it feasible.

**Reply by JR3nard** [reply]
OMG, yes, if you can prioritize that it could be simply huge 🤩.

**Reply by JulienMDST** [reply]
Hi,

Very interesting thread. @mkkellogg, I sent you an email about it, it might be in your spam folder. 

Looking forward to hearing from you!

Have a nice day.

Julien

**Reply by JulienMDST** [reply]
Hello @mkkellogg ,

I hope you're well. Did you well received my email ?

Have a great day 😊.

Julien

**Reply by lxzbg** [reply]
Hi @mkkellogg Based on your work, I simply implemented [a first-order SH that supports the ply model](https://github.com/lxzbg/GaussianSplats3D/tree/test_SH_1). There is a certain quality improvement, but there are still black fog artifacts. Is this because it is not a third-order SH? Could you help me check the correctness of my implementation?

**Reply by mkkellogg** [reply]
@lxzbg I have also been working on my own implementation for 1st degree spherical harmonics. I tried to mimic Aras Pranckevičius' implementation: https://github.com/aras-p/UnityGaussianSplatting. To verify if yours is correct, I would load his implementation in Unity and force it to only use 1st degree SH by commenting out this block: https://github.com/aras-p/UnityGaussianSplatting/blob/0265bd094d67685fa692914c9b9c2f0ce3ba8e98/package/Shaders/GaussianSplatting.hlsl#L154
Then compare the result to yours.

**Comment by mkkellogg** [comment]
So it looks like some of the folks involved with the original gaussian splatting research at INRIA have developed a way to significantly reduce the memory footprint of splat scenes and spherical harmonics: https://repo-sam.inria.fr/fungraph/reduced_3dgs/. As soon as they release their source code I will attempt to port it to my viewer.

**Reply by JR3nard** [reply]
Great to read that Mark. Top !

**Comment by mkkellogg** [comment]
I've been working to support 1st degree spherical harmonics as they are formatted in the original `.ply` format and the result is a mild increase in visual quality. I have a PR up that work here: https://github.com/mkkellogg/GaussianSplats3D/pull/216. I also have a WIP branch to support 2nd degree spherical harmonics here: https://github.com/mkkellogg/GaussianSplats3D/tree/spherical_harmonics_2nd_degree. The visual quality changes are more significant in that branch, so if anyone wants to try it out, let me know how it works for you :)

Note: The `.ply` files currently take a long time to load when incorporating spherical harmonics. Also, you need to pass `sphericalHarmonicsDegree: <desired spherical harmonics degree>` to the viewer to enable it.
